### ISSUE TITLE
FEAT: cross pool transactions

### ISSUE THREAD

**Issue by seghcder** [pmem/issues#988]
Scenario: An application manages multiple discrete pools for performance or NUMA reasons.

Issue: At the moment the library does not support a commit/rollback where data is being updated in two or more pools, eg via nested transactions. Also ref #199, the transactions may not be all committing in the same pool.

Workaround is for the application to manage some form of two phase commit across two or more threads (ref also #321)

**Comment by krzycz** [comment]
It has been discussed a long time ago. The key problem is to guarantee transaction atomicity across multiple pools.  I mean, if the transaction is torn by a power failure, then how to recover from a transaction that has been committed in one pool, but has not been fully committed in the second one?
The old data in undo/redo logs of the first pool could have been already overwritten, so there is no way to roll-back. If we decide to postpone undo/redo log cleanups until the transaction is committed on both sides, then effectively those two pools become one logical pool and must always be open together, as potential recovery in one pool may affect the second one and vice versa.

The problem described in #321 is different. This is when the application (or the library this app links to) is using more than one pool and runs two independent transactions on different pools simultaneously.
If one TX commits, but the second one does not, it does not lead to any serious problem.

**Comment by seghcder** [comment]
My reference to #321 was that I understand it wouldn't be possible for an application to do a two phase commit in the same thread while issue #321 exists. We'd need to co-ordinate between two separate threads. Or is the issue at process level?

If the case is closed on library support for cross pool transactions, then that's also ok :-)

**Comment by pbalcer** [comment]
@seghcder Yes, to properly implement this functionality, we'd need a way for group transaction commit. And now when I think about how the API would look like, we would also need #321,  since there needs to be a way to actually run a second transaction before we do the group commit. We also would need a handle to the transaction (`struct pobj_tx_ctx *`) that we could provide to the group commit function.

@krzycz Right, but let's look at an example:
```
struct root *rootA = ...;
TX_BEGIN(popA, popB) {
  pmemobj_tx_add_range_direct(popA, &rootA->oid, 16);
  rootA->oid = pmemobj_tx_alloc(popB, ...);
} TX_END
```
We just created a one-way reference. To make this work, we would need to synchronize the on_commit phases of those two pools, and only commit if we are 100% sure that both pools will proceed with the commit - just like you say. We would need to forbid cross-pool snapshots though... Each pool need to be responsible for its own logs.
But those two pools CAN be used standalone. In the case of popB, there's no issue. In the case of popA, we will get NULL from `pmemobj_direct`, so the application can account for the fact that a certain amount of data is currently not "connected".

A better question would be: why? The API would look bad...
The only reason I can think of right now is having separate pools for NUMA nodes, so that application can optimally manage and allocate memory in a multi-socket platform.
But that could be solved much easier at an allocator level by just adding a hint to palloc to allocate from a certain address range...

Anything else?

**Comment by seghcder** [comment]
Just adding a few more use potential use cases - 

- Even on a single socket system, having two pool files in the same namespace seems to be faster than one larger one (at least in filesystem based dax). 
- In multi-pool setups, it might allow moving and object from one pool to another within a transaction, eg for re-balancing or if we were moving all objects to a smaller pool and then removing a larger one at runtime. 
- We need to commit a message during a transaction, and we keep the message storage in a separate pool (eg to better manage fragmentation)

**Comment by marcinslusarz** [comment]
1) Can you explain how did you measure this?

2) and 3) look more like a request for dynamic/advanced space management in obj...

3) You can use allocations classes to solve some problems related to fragmentation.
